rse-py-package.Rmd

```diff
@@ -0,0 +1,1477 @@
+        7  07-23-2019 20:30   zipfpy-0.1.dist-info/top_level.txt
+       92  07-23-2019 20:30   zipfpy-0.1.dist-info/WHEEL
+      172  07-23-2019 20:30   zipfpy-0.1.dist-info/METADATA
+      606  07-23-2019 20:30   zipfpy-0.1.dist-info/RECORD
+---------                     -------
+     5649                     8 files
+```
+
+These distribution files can now be distributed via `pypi`, the standard Python
+repository for Python packages.
+
+There is a `testpypi` repository that lets us test putting packages
+on pypi without having them appear on the standard repository; let's
+start with that.  Go to [`https://test.pypi.org`](https://test.pypi.org),
+create an account, and confirm your email.
+
+The preferred tool for uploading packages to the PyPI family of repositories
+is called [`twine`](https://twine.readthedocs.io/en/latest/); we will
+install that package:
+
+```shell
+$ pip install twine
+```
+and now we can upload our distributions:
+
+```shell
+$ twine upload --repository-url https://test.pypi.org/legacy/ dist/*
+Enter your username: pterry
+Enter your passowrd: *********
+Uploading zipfpy-0.1-py3-none-any.whl
+...
+Uploading zipfpy-0.1.tar.gz
+...
+```
+
+The `--repository-url` allows us to specify the test PyPI repository, and we
+have now uploaded both types of distribution, allowing users with different
+environments to use the wheel distribution if their systems allow it, or
+the source distribution otherwise.
+
+We can now test this with a virtual environment and installing with pip:
+
+```shell
+$ virtualenv ~/envs/testpypi
+Using base prefix '/usr/local/Cellar/python/3.7.2_2/Frameworks/Python.framework/Versions/3.7'
+New python executable in /Users/pterry/envs/testpypi/bin/python3.7
+Also creating executable in /Users/pterry/envs/testpypi/bin/python
+sourInstalling setuptools, pip, wheel...
+$ source ~/envs/testpypi/bin/activate
+(testpypi) $ pip install --index-url https://test.pypi.org/simple/ zipfpy
+Looking in indexes: https://test.pypi.org/simple/
+Collecting zipfpy
+  Using cached https://test-files.pythonhosted.org/packages/f8/ac/d152eed95bff7567954becef22b8fe0f2aae15008fbeb346c69f75b9d2a0/zipfpy-0.1-py3-none-any.whl
+Installing collected packages: zipfpy
+Successfully installed zipfpy-0.1
+(testpypi) $
+```
+
+Success!  Once we are happy with how our package appears in testpypi
+(including its [project page](https://test.pypi.org/project/zipfpy/)),
+we can go through the same process to put it on the main pypi repository.
+But first we have a few improvements to make to our setup.py so that
+our package webpage is useful and to make it easier for potential users
+to find, understand, and install our package.
+
+
+### Exercise: Clean up warning messages
+
+-   FIXME: clean up warning messages from `python setup.py sdist`
+
+### Exercise License, longinfo, and classification metadata
+
+-   FIXME: add additional metadata to setup.py
+
+### Exercise: Requirements in setup.py
```

Comment:
Worth mentioning [versioneer](https://github.com/warner/python-versioneer) to automate version strings from git tags?